crates/sui-core/src/checkpoints/mod.rs

```diff
@@ -722,40 +735,14 @@ impl CheckpointStore {
                         .global
                         .checkpoint_items(&diff, proposal.transactions.transactions.clone())
                     {
-                        // Take all certificates and schedule them for execution here.
-                        // We need to at the very least save the certificates that we
-                        // have not executed, to make sure they are available.
-                        let mut batch = self.unprocessed_contents.batch();
-                        batch = batch
-                            .insert_batch(
-                                &self.unprocessed_contents,
-                                reconstructed.extra_transactions,
-                            )
-                            .map_err(|e| FragmentInternalError::Error(e.into()))?;
-                        batch
-                            .write()
-                            .map_err(|e| FragmentInternalError::Error(e.into()))?;
-
                         let contents = CheckpointContents::new(contents.into_iter());
-                        let previous_digest = self
-                            .get_prev_checkpoint_digest(next_sequence_number)
-                            .map_err(FragmentInternalError::Error)?;
-                        let summary = CheckpointSummary::new(
-                            next_sequence_number,
-                            &contents,
-                            previous_digest,
-                        );
-                        self.handle_internal_set_checkpoint(summary, &contents)
-                            .map_err(FragmentInternalError::Error)?;
-                        return Ok(true);
+                        return Ok(Some(contents));
                     }
                 }
             }
-
-            // TODO: here define what we do if we do not have enough info
-            //       to reconstruct the checkpoint. We can store the global waypoints
-            //       and actively wait for someone else to give us the data?
-
+        } else {
```

Comment:
I don't quite understand this else block.
Say the validator just haven't yet received enough (or any) fragments for the new checkpoint, the call to `FragmentReconstruction::construct` would return Ok(None), and we would fall to this else block, which would then set `no_more_fragments` to true?

Reply:
This is a bug indeed. Changing locals should happen in the else clause of the inner if, not the outer one. Fixing and testing.

Reply:
Now we set `no_more_fragments` to true when we don't have a local proposal. What's the rational behind this?